*** /dspace/orig-1.2.2/dspace-1.2.1-source/src/org/dspace/app/webui/servlet/SubmitServlet.java	Wed Dec 22 12:48:35 2004
--- /dspace/orig-1.2.2/dspace-1.2.2-source/src/org/dspace/app/webui/servlet/SubmitServlet.java	Thu Apr 28 16:35:01 2005
***************
*** 1,9 ****
  /*
   * SubmitServlet.java
   *
!  * Version: $Revision: 1.36 $
   *
!  * Date: $Date: 2004/12/22 17:48:35 $
   *
   * Copyright (c) 2002, Hewlett-Packard Company and Massachusetts
   * Institute of Technology.  All rights reserved.
--- 1,9 ----
  /*
   * SubmitServlet.java
   *
!  * Version: $Revision: 1.37.2.1 $
   *
!  * Date: $Date: 2005/04/28 20:35:01 $
   *
   * Copyright (c) 2002, Hewlett-Packard Company and Massachusetts
   * Institute of Technology.  All rights reserved.
***************
*** 47,52 ****
--- 47,53 ----
  import java.sql.SQLException;
  import java.util.LinkedList;
  import java.util.List;
+ import java.util.ArrayList;
  
  import javax.servlet.ServletException;
  import javax.servlet.http.HttpServletRequest;
***************
*** 53,58 ****
--- 54,62 ----
  import javax.servlet.http.HttpServletResponse;
  
  import org.apache.log4j.Logger;
+ 
+ import org.dspace.app.webui.util.DCInputsReader;
+ import org.dspace.app.webui.util.DCInput;
  import org.dspace.app.webui.util.FileUploadRequest;
  import org.dspace.app.webui.util.JSPManager;
  import org.dspace.app.webui.util.SubmissionInfo;
***************
*** 121,127 ****
   * the submission forms to render the user's progress through the process.
   * 
   * @author Robert Tansley
!  * @version $Revision: 1.36 $
   */
  public class SubmitServlet extends DSpaceServlet
  {
--- 125,131 ----
   * the submission forms to render the user's progress through the process.
   * 
   * @author Robert Tansley
!  * @version $Revision: 1.37.2.1 $
   */
  public class SubmitServlet extends DSpaceServlet
  {
***************
*** 133,143 ****
      /** Ask initial questions about the submission step */
      public static final int INITIAL_QUESTIONS = 1;
  
!     /** Edit DC metadata page 1 step */
      public static final int EDIT_METADATA_1 = 2;
  
!     /** Edit DC metadata page 2 step */
!     public static final int EDIT_METADATA_2 = 3;
  
      /**
       * Upload files step. Note this doesn't correspond to an actual page, since
--- 137,152 ----
      /** Ask initial questions about the submission step */
      public static final int INITIAL_QUESTIONS = 1;
  
!     /** Edit DC metadata first page step */
      public static final int EDIT_METADATA_1 = 2;
  
!     /** Edit DC metadata last page step 
!      * Current value allows up to 6 distinct pages.
!      * If more are needed, renumber this value and those 
!      * that follow it (up to SUBMISSION_COMPLETE), but
!      * note that the progress bar will stretch badly beyond 6.
!      */
!     public static final int EDIT_METADATA_2 = 7;
  
      /**
       * Upload files step. Note this doesn't correspond to an actual page, since
***************
*** 144,163 ****
       * the upload file step consists of a number of pages with no definite
       * order. This is just used for the progress bar.
       */
!     public static final int UPLOAD_FILES = 4;
  
      /** Review submission step */
!     public static final int REVIEW_SUBMISSION = 5;
  
      /** optional CC license step */
!     public static final int CC_LICENSE = 6;
  
      /** Grant (deposit) license step */
!     public static final int GRANT_LICENSE = 7;
  
      /** Submission completed step */
!     public static final int SUBMISSION_COMPLETE = 8;
! 
      // Steps which aren't part of the main sequence, but rather
      // short "diversions" are given high step numbers. The main sequence
      // is defined as being steps 0 to SUBMISSION_COMPLETE.
--- 153,172 ----
       * the upload file step consists of a number of pages with no definite
       * order. This is just used for the progress bar.
       */
!     public static final int UPLOAD_FILES = 8;
  
      /** Review submission step */
!     public static final int REVIEW_SUBMISSION = 9;
  
      /** optional CC license step */
!     public static final int CC_LICENSE = 10;
  
      /** Grant (deposit) license step */
!     public static final int GRANT_LICENSE = 11;
  
      /** Submission completed step */
!     public static final int SUBMISSION_COMPLETE = 12;
!     
      // Steps which aren't part of the main sequence, but rather
      // short "diversions" are given high step numbers. The main sequence
      // is defined as being steps 0 to SUBMISSION_COMPLETE.
***************
*** 188,193 ****
--- 197,212 ----
  
      /** log4j logger */
      private static Logger log = Logger.getLogger(SubmitServlet.class);
+     
+     /** hash of all submission forms details */
+     private DCInputsReader inputsReader;
+     
+     public SubmitServlet()
+     	throws ServletException
+     {
+ 	// read configurable submissions forms data
+ 	inputsReader = new DCInputsReader();
+     }
  
      protected void doDSGet(Context context, HttpServletRequest request,
              HttpServletResponse response) throws ServletException, IOException,
***************
*** 357,362 ****
--- 376,387 ----
                  step = -1;
              }
          }
+         
+         if (step >= EDIT_METADATA_1 && step <= EDIT_METADATA_2)
+         {
+         	processEditMetadata(context, request, response, subInfo, step);
+         	return;
+         }
  
          switch (step)
          {
***************
*** 364,380 ****
              processInitialQuestions(context, request, response, subInfo);
  
              break;
! 
!         case EDIT_METADATA_1:
!             processEditMetadataOne(context, request, response, subInfo);
! 
!             break;
! 
!         case EDIT_METADATA_2:
!             processEditMetadataTwo(context, request, response, subInfo);
! 
!             break;
! 
          case GET_FILE_FORMAT:
              processGetFileFormat(context, request, response, subInfo);
  
--- 389,395 ----
              processInitialQuestions(context, request, response, subInfo);
  
              break;
!   
          case GET_FILE_FORMAT:
              processGetFileFormat(context, request, response, subInfo);
  
***************
*** 756,762 ****
      }
  
      /**
!      * process input from edit-metadata-1.jsp
       * 
       * @param context
       *            current DSpace context
--- 771,777 ----
      }
  
      /**
!      * process input from edit-metadata.jsp
       * 
       * @param context
       *            current DSpace context
***************
*** 766,776 ****
       *            current servlet response object
       * @param subInfo
       *            submission info object
       */
!     private void processEditMetadataOne(Context context,
!             HttpServletRequest request, HttpServletResponse response,
!             SubmissionInfo subInfo) throws ServletException, IOException,
!             SQLException, AuthorizeException
      {
          String buttonPressed = UIUtil.getSubmitButton(request, "submit_next");
  
--- 781,795 ----
       *            current servlet response object
       * @param subInfo
       *            submission info object
+      * @param curStep
+      *            page number in edit sequence, starting at zero
       */
!     private void processEditMetadata(Context context,
!         HttpServletRequest request,
!         HttpServletResponse response,
!         SubmissionInfo subInfo,
! 		int curStep)
!         throws ServletException, IOException, SQLException, AuthorizeException
      {
          String buttonPressed = UIUtil.getSubmitButton(request, "submit_next");
  
***************
*** 777,920 ****
          // Firstly, check for a click of the cancel button.
          if (buttonPressed.equals("submit_cancel"))
          {
!             doCancellation(request, response, subInfo, EDIT_METADATA_1,
!                     EDIT_METADATA_1);
! 
              return;
          }
  
          Item item = subInfo.submission.getItem();
! 
!         // Update the item metadata.
!         readNames(request, item, "contributor", "author", true);
!         readText(request, item, "title", null, false, "en");
! 
!         // verify that a title has been entered
!         DCValue[] dcValues = item.getDC("title", null, "en");
!         boolean titleMissing = dcValues.length == 0;
! 
!         if (subInfo.submission.hasMultipleTitles())
          {
!             readText(request, item, "title", "alternative", true, "en");
          }
  
!         readSeriesNumbers(request, item, "relation", "ispartofseries", true);
  
!         // Type
!         item.clearDC("type", null, Item.ANY);
  
!         String[] types = request.getParameterValues("type");
! 
!         if (types != null)
!         {
!             for (int i = 0; i < types.length; i++)
!             {
!                 item.addDC("type", null, "en", types[i]);
!             }
!         }
! 
!         // FIXME: Maybe should do integrity check using language object
!         readText(request, item, "language", "iso", false, null);
! 
!         // Identifiers - A special case so do this here
!         // First, remove existing identifiers.
!         subInfo.submission.getItem().clearDC("identifier", Item.ANY, Item.ANY);
! 
!         // Get qualifiers and values
!         List quals = getRepeatedParameter(request, "identifier_qualifier");
!         List vals = getRepeatedParameter(request, "identifier_value");
! 
!         // Add to item DC
!         for (int i = 0; i < vals.size(); i++)
!         {
!             String thisQual = (String) quals.get(i);
!             String thisVal = (String) vals.get(i);
! 
!             if (!buttonPressed.equals("submit_identifier_remove_" + i)
!                     && !thisVal.equals(""))
!             {
!                 item.addDC("identifier", thisQual, null, thisVal);
!             }
!         }
! 
!         boolean issueDateMissing = false;
! 
!         if (subInfo.submission.isPublishedBefore())
!         {
!             readDate(request, item, "date", "issued");
! 
!             // check that at least year has been entered - flag if not
!             DCValue[] dateArray = item.getDC("date", "issued", Item.ANY);
!             DCDate dateIssued = new DCDate(
!                     ((dateArray.length > 0) ? dateArray[0].value : ""));
! 
!             if (dateIssued.getYear() <= 0)
!             {
!                 issueDateMissing = true;
!             }
! 
!             // Careful that this happens AFTER identifier.* is blown away
!             // above!!
!             readText(request, item, "identifier", "citation", false, "en");
!             readText(request, item, "publisher", null, false, "en");
!         }
! 
          int nextStep = -1;
! 
!         // Proceed according to button pressed
!         if (buttonPressed.equals("submit_contributor_author_more"))
          {
!             // "Add more" clicked for authors
!             subInfo.moreBoxesFor = "contributor_author";
!             subInfo.jumpToField = "contributor_author";
!             nextStep = EDIT_METADATA_1;
          }
!         else if (subInfo.submission.hasMultipleTitles()
!                 && buttonPressed.equals("submit_title_alternative_more"))
          {
!             // "Add more" clicked for alternative titles
!             subInfo.moreBoxesFor = "title_alternative";
!             subInfo.jumpToField = "title_alternative";
!             nextStep = EDIT_METADATA_1;
          }
-         else if (buttonPressed.equals("submit_relation_ispartofseries_more"))
-         {
-             // "Add more" clicked for series/number
-             subInfo.moreBoxesFor = "relation_ispartofseries";
-             subInfo.jumpToField = "relation_ispartofseries";
-             nextStep = EDIT_METADATA_1;
-         }
-         else if (buttonPressed.equals("submit_identifier_more"))
-         {
-             // "Add more" clicked for identifier
-             subInfo.moreBoxesFor = "identifier";
-             subInfo.jumpToField = "identifier";
-             nextStep = EDIT_METADATA_1;
-         }
-         else if (buttonPressed.equals("submit_prev"))
-         {
-             nextStep = INITIAL_QUESTIONS;
-         }
          else if (buttonPressed.equals("submit_next"))
          {
!             // return to edit metadata 1 if title or issue date missing
!             if (titleMissing || issueDateMissing)
!             {
!                 subInfo.missing = true;
!                 subInfo.jumpToField = (titleMissing ? "title"
!                         : "date_issued_year");
!                 nextStep = EDIT_METADATA_1;
!             }
!             else
!             {
!                 userHasReached(subInfo, EDIT_METADATA_2);
!                 nextStep = EDIT_METADATA_2;
!             }
          }
          else if (buttonPressed.indexOf("remove") > -1)
          {
              // Remove button pressed - stay with same form
!             nextStep = EDIT_METADATA_1;
          }
  
          // Write changes to database
--- 796,964 ----
          // Firstly, check for a click of the cancel button.
          if (buttonPressed.equals("submit_cancel"))
          {
!             doCancellation(request, response, subInfo, curStep, curStep);
              return;
          }
  
          Item item = subInfo.submission.getItem();
!         
!         // lookup applicable inputs
!         Collection c = subInfo.submission.getCollection();
!         DCInput[] inputs = inputsReader.getInputs(c.getHandle()).getPageRows(curStep - EDIT_METADATA_1,
! 									 subInfo.submission.hasMultipleTitles(),
! 									 subInfo.submission.isPublishedBefore());
!  
!         // clear out all item metadata defined for this collection and page
!         for (int i = 0; i < inputs.length; i++)
          {
!         	String dcQualifier = inputs[i].getQualifier();
!         	if (dcQualifier == null && inputs[i].getInputType().equals("qualdrop_value"))
!         	{
!         		dcQualifier = Item.ANY;
!         	}
!         	item.clearDC(inputs[i].getElement(), dcQualifier, Item.ANY);
          }
  
!         // now update the item metadata.
!       	String fieldName;
!         boolean moreInput = false;
!     	for (int j = 0; j < inputs.length; j++)
!     	{
!     	   String dcElement = inputs[j].getElement();
!            String dcQualifier = inputs[j].getQualifier();
!     	   if (dcQualifier != null && ! dcQualifier.equals(Item.ANY))
!     	   {
!     	   		fieldName = dcElement + '_' + dcQualifier;
!     	   }
!     	   else
!     	   {
!     	   		fieldName = dcElement;
!     	   }
  
!     	   String inputType = inputs[j].getInputType();
!     	   if (inputType.equals("name"))
!     	   {
!     	   		readNames(request, item, dcElement, dcQualifier, 
!     	   				  inputs[j].getRepeatable());
!     	   }
!     	   else if (inputType.equals("date"))
!     	   {
!     	   		readDate(request, item, dcElement, dcQualifier);
!     	   }
!     	   else if (inputType.equals("series"))
!     	   {
!     	   		readSeriesNumbers(request, item, dcElement, dcQualifier, 
!     	   						  inputs[j].getRepeatable());
!     	   }
!     	   else if (inputType.equals("qualdrop_value"))
!     	   {
!     	      List quals = getRepeatedParameter(request, dcElement + "_qualifier");
!     	      List vals = getRepeatedParameter(request, dcElement + "_value");
!     	      for (int z = 0; z < vals.size(); z++)
!     	      {
!     	      		String thisQual = (String)quals.get(z);
!     	      		if ( "".equals(thisQual) )
!     	      		{
!     	      		    thisQual = null;
!     	      		}
!     	      		String thisVal = (String)vals.get(z);
!     	      		if (! buttonPressed.equals("submit_" + dcElement + "_remove_" + z) &&
!     	      			! thisVal.equals(""))
!     	      		{
!     	      			item.addDC(dcElement, thisQual, null, thisVal);
!     	      		}
!     	      }
!     	   }
!     	   else if (inputType.equals("dropdown"))
!     	   {
!     	      String[] vals = request.getParameterValues(fieldName);
!     	      if (vals != null)
!     	      {
!     	      	for (int z = 0; z < vals.length; z++)
!     	      	{
!     	      		item.addDC(dcElement, dcQualifier, "en", vals[z]);
!     	      	}
!     	      }
!     	   }
!     	   else if ((inputType.equals("onebox")) || (inputType.equals("twobox")) || 
!     	   			(inputType.equals("textarea")))
!     	   {
!     	   		readText(request, item, dcElement, dcQualifier, 
!     	   				 inputs[j].getRepeatable(), "en");
!     	   }
!     	   else 
!     	   {
!     	       throw new ServletException("Field "+ fieldName + 
!     	       		                      " has an unknown input type: " + inputType);
!     	   }
  
!            // Proceed according to button pressed
!     	   if (! moreInput && buttonPressed.equals("submit_" + fieldName + "_more"))
!     	   {
!     	      subInfo.moreBoxesFor = fieldName;
!     	      subInfo.jumpToField = fieldName;
!     	      moreInput = true;
!     	   }
!     	}
!     	
          int nextStep = -1;
!         if ( moreInput )
          {
!         	nextStep = curStep;
          }
!     	
!         if (buttonPressed.equals("submit_prev"))
          {
!             // NB: code here assumes steps are sequentially numbered!
!             nextStep = curStep-1;
          }
          else if (buttonPressed.equals("submit_next"))
          {
!         	// scan for missing fields
!         	String gotoField = null;
!         	subInfo.missingFields = new ArrayList();
!         	//subInfo.missingRowNums = new Vector();
!         	for (int i = 0; i < inputs.length; i++)
!         	{
!         	    String element = inputs[i].getElement();
!         	    String qual = inputs[i].getQualifier();
!         	    DCValue[] valArray = item.getDC(element, qual, Item.ANY);
!         	    boolean isEmpty = (valArray.length == 0);
!         	    if (inputs[i].isRequired() && isEmpty)
!         	    {
!         	    	subInfo.missingFields.add( new Integer(i) );
!         	    	if (qual != null && !qual.equals("*"))
!         	    	{
!         	    		gotoField = element + '_' + qual;
!         	    	}
!         	    	else
!         	    	{
!         	    		gotoField = element;
!         	    	}
!         	    }
!         	}
!         	// return to current edit metadata screen if any fields missing
!         	if (subInfo.missingFields.size() > 0 )
!         	{
!         		subInfo.jumpToField = gotoField;
!         		nextStep = curStep;
!         	}
!         	else
!         	{
!         		// determine next step - skipping unused MD pages
!         		int lastMDPage = EDIT_METADATA_1 + inputsReader.getNumberInputPages(c.getHandle()) - 1;
!         		if ( curStep == lastMDPage )
!         		{
!         		    curStep = EDIT_METADATA_2;
!         		}
!         		userHasReached(subInfo, curStep+1);
!         		nextStep = curStep+1;
!         	}
          }
          else if (buttonPressed.indexOf("remove") > -1)
          {
              // Remove button pressed - stay with same form
!             nextStep = curStep;
          }
  
          // Write changes to database
***************
*** 933,1014 ****
      }
  
      /**
-      * process input from edit-metadata-2.jsp
-      * 
-      * @param context
-      *            current DSpace context
-      * @param request
-      *            current servlet request object
-      * @param response
-      *            current servlet response object
-      * @param subInfo
-      *            submission info object
-      */
-     private void processEditMetadataTwo(Context context,
-             HttpServletRequest request, HttpServletResponse response,
-             SubmissionInfo subInfo) throws ServletException, IOException,
-             SQLException, AuthorizeException
-     {
-         String buttonPressed = UIUtil.getSubmitButton(request, "submit_next");
- 
-         // Firstly, check for a click of the cancel button.
-         if (buttonPressed.equals("submit_cancel"))
-         {
-             doCancellation(request, response, subInfo, EDIT_METADATA_2,
-                     EDIT_METADATA_2);
- 
-             return;
-         }
- 
-         Item item = subInfo.submission.getItem();
- 
-         // Update object from form values
-         readText(request, item, "subject", null, true, "en");
-         readText(request, item, "description", "abstract", false, "en");
-         readText(request, item, "description", "sponsorship", false, "en");
-         readText(request, item, "description", null, false, "en");
- 
-         // Proceed according to button pressed
-         int nextStep = -1;
- 
-         if (buttonPressed.equals("submit_subject_more"))
-         {
-             // "Add more" clicked for subject keywords
-             subInfo.moreBoxesFor = "subject";
-             subInfo.jumpToField = "subject";
-             nextStep = EDIT_METADATA_2;
-         }
-         else if (buttonPressed.equals("submit_prev"))
-         {
-             nextStep = EDIT_METADATA_1;
-         }
-         else if (buttonPressed.equals("submit_next"))
-         {
-             userHasReached(subInfo, UPLOAD_FILES);
-             nextStep = UPLOAD_FILES;
-         }
-         else if (buttonPressed.indexOf("remove") > -1)
-         {
-             // Remove button pressed - stay with same form
-             nextStep = EDIT_METADATA_2;
-         }
- 
-         // Write changes to database
-         subInfo.submission.update();
- 
-         if (nextStep != -1)
-         {
-             doStep(context, request, response, subInfo, nextStep);
-         }
-         else
-         {
-             doStepJump(context, request, response, subInfo);
-         }
- 
-         context.complete();
-     }
- 
-     /**
       * Process the input from the choose file page
       * 
       * @param context
--- 977,982 ----
***************
*** 1132,1139 ****
              }
              else
              {
!                 // No files, go back to edit metadata 2
!                 doStep(context, request, response, subInfo, EDIT_METADATA_2);
              }
          }
          else if (buttonPressed.equals("submit_next"))
--- 1100,1110 ----
              }
              else
              {
!                 // No files, go back to last edit metadata page
!             	Collection c = subInfo.submission.getCollection();
!             	int lastPage = EDIT_METADATA_1 + 
! 				               inputsReader.getNumberInputPages( c.getHandle() ) - 1;
!                 doStep(context, request, response, subInfo, lastPage);
              }
          }
          else if (buttonPressed.equals("submit_next"))
***************
*** 1159,1167 ****
              {
                  // If we get here, there was a problem uploading, but we
                  // still know which submission we're dealing with
!                 request.setAttribute("submission.info", subInfo);
!                 JSPManager.showJSP(request, response,
!                         "/submit/upload-error.jsp");
              }
          }
          else
--- 1130,1137 ----
              {
                  // If we get here, there was a problem uploading, but we
                  // still know which submission we're dealing with
!             	showProgressAwareJSP(request, response,	subInfo,
!             	 	  	     "/submit/upload-error.jsp");
              }
          }
          else
***************
*** 1262,1269 ****
          {
              // In some cases, this might be expected to go back
              // to the "choose file" page, but that doesn't make
!             // a great deal of sense, so go back to edit metadata 2.
!             doStep(context, request, response, subInfo, EDIT_METADATA_2);
          }
          else if (buttonPressed.equals("submit_next"))
          {
--- 1232,1242 ----
          {
              // In some cases, this might be expected to go back
              // to the "choose file" page, but that doesn't make
!             // a great deal of sense, so go back to last edit metadata page.
!         	Collection c = subInfo.submission.getCollection();
!         	int lastPage = EDIT_METADATA_1 + 
! 			               inputsReader.getNumberInputPages( c.getHandle() ) - 1;
!             doStep(context, request, response, subInfo, lastPage);
          }
          else if (buttonPressed.equals("submit_next"))
          {
***************
*** 1295,1302 ****
              }
  
              // Upload another file
!             request.setAttribute("submission.info", subInfo);
!             JSPManager.showJSP(request, response, "/submit/choose-file.jsp");
          }
          else if (buttonPressed.equals("submit_show_checksums"))
          {
--- 1268,1274 ----
              }
  
              // Upload another file
!             showProgressAwareJSP(request, response, subInfo, "/submit/choose-file.jsp");
          }
          else if (buttonPressed.equals("submit_show_checksums"))
          {
***************
*** 1331,1339 ****
  
              // Display the form letting them change the description
              subInfo.bitstream = bitstream;
!             request.setAttribute("submission.info", subInfo);
!             JSPManager.showJSP(request, response,
!                     "/submit/change-file-description.jsp");
          }
          else if (buttonPressed.startsWith("submit_remove_"))
          {
--- 1303,1310 ----
  
              // Display the form letting them change the description
              subInfo.bitstream = bitstream;
!             showProgressAwareJSP(request, response, subInfo,
!                 		 "/submit/change-file-description.jsp");
          }
          else if (buttonPressed.startsWith("submit_remove_"))
          {
***************
*** 1436,1443 ****
          // no real options on the page, just retry!
          if (buttonPressed.equals("submit"))
          {
!             request.setAttribute("submission.info", subInfo);
!             JSPManager.showJSP(request, response, "/submit/choose-file.jsp");
          }
          else
          {
--- 1407,1414 ----
          // no real options on the page, just retry!
          if (buttonPressed.equals("submit"))
          {
!             showProgressAwareJSP(request, response, subInfo,
!             		             "/submit/choose-file.jsp");
          }
          else
          {
***************
*** 1633,1640 ****
              WorkflowManager.start(context, (WorkspaceItem) subInfo.submission);
  
              // FIXME: pass in more information about what happens next?
!             JSPManager.showJSP(request, response, "/submit/complete.jsp");
! 
              context.complete();
          }
          else if (request.getParameter("submit_reject") != null)
--- 1604,1610 ----
              WorkflowManager.start(context, (WorkspaceItem) subInfo.submission);
  
              // FIXME: pass in more information about what happens next?
!             showProgressAwareJSP(request, response, subInfo, "/submit/complete.jsp");
              context.complete();
          }
          else if (request.getParameter("submit_reject") != null)
***************
*** 1705,1711 ****
              // RLR hack - need to distinguish between progress bar real
              // submission
              String ccLicenseUrl = request.getParameter("cc_license_url");
- 
              if ((ccLicenseUrl != null) && (ccLicenseUrl.length() > 0))
              {
                  Item item = subInfo.submission.getItem();
--- 1675,1680 ----
***************
*** 1826,1854 ****
              HttpServletResponse response, SubmissionInfo subInfo, int step)
              throws ServletException, IOException, SQLException
      {
!         // All steps require the submitforminfo
!         request.setAttribute("submission.info", subInfo);
  
          switch (step)
          {
          case INITIAL_QUESTIONS:
!             JSPManager.showJSP(request, response,
!                     "/submit/initial-questions.jsp");
! 
              break;
! 
!         case EDIT_METADATA_1:
!             JSPManager
!                     .showJSP(request, response, "/submit/edit-metadata-1.jsp");
! 
!             break;
! 
!         case EDIT_METADATA_2:
!             JSPManager
!                     .showJSP(request, response, "/submit/edit-metadata-2.jsp");
! 
!             break;
! 
          case UPLOAD_FILES:
              showFirstUploadPage(context, request, response, subInfo);
  
--- 1795,1825 ----
              HttpServletResponse response, SubmissionInfo subInfo, int step)
              throws ServletException, IOException, SQLException
      {
!         // determine collection
!         Collection c = subInfo.submission.getCollection();
!         
!         if ( step >= EDIT_METADATA_1 && step <= EDIT_METADATA_2 )
!         {
!         	// requires configurable form info per collection
!         	request.setAttribute( "submission.inputs", inputsReader.getInputs(c.getHandle()));
!         	// also indicate page
!            	request.setAttribute( "submission.page", new Integer(step) );
! 		showProgressAwareJSP(request, response, subInfo,
!                   		     "/submit/edit-metadata.jsp");
!             return;
!         }
  
          switch (step)
          {
          case INITIAL_QUESTIONS:
!             // requires configurable form info per collection
!             request.setAttribute( "submission.inputs", inputsReader.getInputs(c.getHandle()));
!             showProgressAwareJSP(request, response, subInfo,
!                 		 "/submit/initial-questions.jsp");
              break;
!             
!         /* EDIT_METADATA cases handled above */
!             
          case UPLOAD_FILES:
              showFirstUploadPage(context, request, response, subInfo);
  
***************
*** 1855,1880 ****
              break;
  
          case CHOOSE_FILE:
!             JSPManager.showJSP(request, response, "/submit/choose-file.jsp");
! 
              break;
  
          case FILE_LIST:
              showUploadFileList(request, response, subInfo, false, false);
- 
              break;
  
          case REVIEW_SUBMISSION:
!             JSPManager.showJSP(request, response, "/submit/review.jsp");
! 
              break;
  
          case GRANT_LICENSE:
- 
-             Collection c = subInfo.submission.getCollection();
              request.setAttribute("license", c.getLicense());
!             JSPManager.showJSP(request, response, "/submit/show-license.jsp");
! 
              break;
  
          case CC_LICENSE:
--- 1826,1850 ----
              break;
  
          case CHOOSE_FILE:
!             showProgressAwareJSP(request, response, subInfo,
!             		         "/submit/choose-file.jsp");
              break;
  
          case FILE_LIST:
              showUploadFileList(request, response, subInfo, false, false);
              break;
  
          case REVIEW_SUBMISSION:
!             // requires configurable form info per collection
!             request.setAttribute( "submission.inputs", inputsReader.getInputs(c.getHandle()));
!             showProgressAwareJSP(request, response, subInfo,
!             		         "/submit/review.jsp");
              break;
  
          case GRANT_LICENSE:
              request.setAttribute("license", c.getLicense());
!             showProgressAwareJSP(request, response, subInfo,
!             		         "/submit/show-license.jsp");
              break;
  
          case CC_LICENSE:
***************
*** 1881,1896 ****
  
              // Do we already have a CC license?
              Item item = subInfo.submission.getItem();
!             boolean exists = CreativeCommons.hasLicense(context, item);
!             request.setAttribute("cclicense.exists", new Boolean(exists));
!             JSPManager.showJSP(request, response,
!                     "/submit/creative-commons.jsp");
! 
              break;
  
          case SUBMISSION_COMPLETE:
!             JSPManager.showJSP(request, response, "/submit/complete.jsp");
! 
              break;
  
          default:
--- 1851,1865 ----
  
              // Do we already have a CC license?
              Item item = subInfo.submission.getItem();
!             boolean exists =  CreativeCommons.hasLicense(context, item);
!             request.setAttribute("cclicense.exists", new Boolean(exists) );
!             showProgressAwareJSP(request, response, subInfo,
!             		         "/submit/creative-commons.jsp");
              break;
  
          case SUBMISSION_COMPLETE:
!             showProgressAwareJSP(request, response, subInfo,
!             		         "/submit/complete.jsp");
              break;
  
          default:
***************
*** 1927,1936 ****
          }
          else
          {
-             request.setAttribute("submission.info", subInfo);
              request.setAttribute("step", String.valueOf(step));
              request.setAttribute("display.step", String.valueOf(displayStep));
!             JSPManager.showJSP(request, response, "/submit/cancel.jsp");
          }
      }
  
--- 1896,1904 ----
          }
          else
          {
              request.setAttribute("step", String.valueOf(step));
              request.setAttribute("display.step", String.valueOf(displayStep));
!             showProgressAwareJSP(request, response, subInfo, "/submit/cancel.jsp");
          }
      }
  
***************
*** 1987,1993 ****
              ServletException, IOException
      {
          // Set required attributes
-         request.setAttribute("submission.info", subInfo);
          request.setAttribute("just.uploaded", new Boolean(justUploaded));
          request.setAttribute("show.checksums", new Boolean(showChecksums));
  
--- 1955,1960 ----
***************
*** 1994,2007 ****
          // Always go to advanced view in workflow mode
          if (isWorkflow(subInfo) || subInfo.submission.hasMultipleFiles())
          {
!             JSPManager.showJSP(request, response,
!                     "/submit/upload-file-list.jsp");
          }
          else
          {
              // FIXME: Assume one and only one bitstream
!             JSPManager.showJSP(request, response,
!                     "/submit/show-uploaded-file.jsp");
          }
      }
  
--- 1961,1974 ----
          // Always go to advanced view in workflow mode
          if (isWorkflow(subInfo) || subInfo.submission.hasMultipleFiles())
          {
!             showProgressAwareJSP(request, response, subInfo,
!        				 "/submit/upload-file-list.jsp");
          }
          else
          {
              // FIXME: Assume one and only one bitstream
!             showProgressAwareJSP(request, response, subInfo,
!        				 "/submit/show-uploaded-file.jsp");
          }
      }
  
***************
*** 2028,2036 ****
  
          subInfo.bitstream = bitstream;
  
-         request.setAttribute("submission.info", subInfo);
          request.setAttribute("bitstream.formats", formats);
! 
          // What does the system think it is?
          BitstreamFormat guess = FormatIdentifier
                  .guessFormat(context, bitstream);
--- 1995,2002 ----
  
          subInfo.bitstream = bitstream;
  
          request.setAttribute("bitstream.formats", formats);
!  
          // What does the system think it is?
          BitstreamFormat guess = FormatIdentifier
                  .guessFormat(context, bitstream);
***************
*** 2037,2043 ****
  
          request.setAttribute("guessed.format", guess);
  
!         JSPManager.showJSP(request, response, "/submit/get-file-format.jsp");
      }
  
      //****************************************************************
--- 2003,2010 ----
  
          request.setAttribute("guessed.format", guess);
  
!         showProgressAwareJSP(request, response, subInfo,
!        		             "/submit/get-file-format.jsp");
      }
  
      //****************************************************************
***************
*** 2045,2050 ****
--- 2012,2040 ----
      //             MISCELLANEOUS CONVENIENCE METHODS
      //****************************************************************
      //****************************************************************
+     
+     /**
+      * Show a JSP after setting attributes needed by progress bar
+      * @param request	the request object
+      * @param response  the response object
+      * @param subInfo   the SubmissionInfo object
+      * @param jspPath	relative path to JSP
+      */
+      private void showProgressAwareJSP(
+             HttpServletRequest request,
+ 	        HttpServletResponse response,
+ 	        SubmissionInfo subInfo,
+ 			String jspPath)
+      		throws ServletException, IOException
+      {
+      	// all JSPs displaying the progress bar need to know the
+      	// number of metadata edit pages
+         subInfo.numMetadataPages =
+         	inputsReader.getNumberInputPages(subInfo.submission.getCollection().getHandle());
+         request.setAttribute("submission.info", subInfo);
+         
+         JSPManager.showJSP(request, response, jspPath);
+      }
  
      /**
       * Get a filled-out submission info object from the parameters in the
***************
*** 2631,2634 ****
  
          return vals;
      }
! }
--- 2621,2624 ----
  
          return vals;
      }
! }
