*** /dspace/orig-1.2.1/dspace-1.2-source/src/org/dspace/handle/HandleManager.java	Mon Apr  7 13:37:50 2003
--- /dspace/dev-1.2/dspace-1.2-source/src/org/dspace/handle/HandleManager.java	Fri Mar 26 14:31:53 2004
***************
*** 38,43 ****
--- 38,56 ----
   * TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
   * USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
   * DAMAGE.
+  *
+  * <pre>
+  * Revision History
+  *
+  *   2004/03/26: Ben
+  *     - make getHandlesForPrefix public
+  *     - fix sql in getHandlesForPrefix
+  *
+  *   2004/03/25: Ben
+  *     - modified getCanonicalForm to use the handle.http-proxy config
+  *       variable
+  *
+  * </pre>
   */
  
  package org.dspace.handle;
***************
*** 112,118 ****
      public static String getCanonicalForm(String handle)
      {
  //        return "hdl:" + handle;
!         return "http://hdl.handle.net/" + handle;
      }
  
  
--- 125,134 ----
      public static String getCanonicalForm(String handle)
      {
  //        return "hdl:" + handle;
! 		  String http_proxy = ConfigurationManager.getProperty("handle.http-proxy");
! 
! 
!         return http_proxy + handle;
      }
  
  
***************
*** 276,286 ****
       * is a String.
       * @exception SQLException If a database error occurs
       */
!     static List getHandlesForPrefix(Context context,
                                             String prefix)
          throws SQLException
      {
!         String sql = "SELECT handle FROM handle WHERE handle LIKE " + prefix + "%";
          TableRowIterator iterator = DatabaseManager.query(context, null, sql);
          List results = new ArrayList();
          while (iterator.hasNext())
--- 292,302 ----
       * is a String.
       * @exception SQLException If a database error occurs
       */
!     public static List getHandlesForPrefix(Context context,
                                             String prefix)
          throws SQLException
      {
!         String sql = "SELECT handle FROM handle WHERE handle LIKE '" + prefix + "%'";
          TableRowIterator iterator = DatabaseManager.query(context, null, sql);
          List results = new ArrayList();
          while (iterator.hasNext())
