*** /dspace/orig-1.2.1/dspace-1.2-source/src/org/dspace/app/webui/util/Authenticate.java	Mon Feb 23 13:32:44 2004
--- /dspace/dev-1.2/dspace-1.2-source/src/org/dspace/app/webui/util/Authenticate.java	Wed Dec 22 10:50:15 2004
***************
*** 65,70 ****
--- 65,83 ----
   */
  public class Authenticate
  {
+    /**
+     * <pre>
+     * Revision History:
+     *
+     *   2004/12/22: Ben
+     *     - change the out List to the actual ldap object
+     *
+     *   2004/12/20: Ben
+     *   - remove ldap ou information when logged out
+     *
+     * </pre>
+     */
+ 
      /** log4j category */
      private static Logger log = Logger.getLogger(Authenticate.class);
  
***************
*** 276,280 ****
--- 289,294 ----
          context.setCurrentUser(null);
          request.removeAttribute("dspace.current.user");
          session.removeAttribute("dspace.current.user.id");
+ 	session.removeAttribute("dspace.current.user.ldap");
      }
  }
