*** /dspace/orig-1.2.1/dspace-1.2-source/src/org/dspace/app/oai/DSpaceOAICatalog.java	Tue Jun 22 11:35:51 2004
--- /dspace/dev-1.2/dspace-1.2-source/src/org/dspace/app/oai/DSpaceOAICatalog.java	Mon Oct 11 14:02:51 2004
***************
*** 36,41 ****
--- 36,51 ----
   * TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
   * USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
   * DAMAGE.
+  *
+  * <pre>
+  * Revision History:
+  * 
+  *   2004/07/14: Ben
+  *     - correctly handle errors parsing the resumption token
+  *
+  *   2004/07/13: Ben
+  *     - add xml escaping
+  * </pre>
   */
  
  package org.dspace.app.oai;
***************
*** 63,68 ****
--- 73,79 ----
  import ORG.oclc.oai.server.verb.NoSetHierarchyException;
  import ORG.oclc.oai.server.verb.OAIInternalServerError;
  
+ import org.dspace.app.oai.OAIDCCrosswalk;
  import org.dspace.content.Collection;
  import org.dspace.content.DSpaceObject;
  import org.dspace.core.ConfigurationManager;
***************
*** 625,631 ****
              {
  				String spec = "<set><setSpec>hdl_" +
  					allCols[i].getHandle().replace('/','_') +
! 					"</setSpec><setName>" + allCols[i].getMetadata("name") +
                      "</setName></set>";
                  
                  sets.add(spec);
--- 636,642 ----
              {
  				String spec = "<set><setSpec>hdl_" +
  					allCols[i].getHandle().replace('/','_') +
! 					"</setSpec><setName>" + OAIDCCrosswalk.escapeXml(allCols[i].getMetadata("name")) +
                      "</setName></set>";
                  
                  sets.add(spec);
