*** /dspace/orig-1.2.2/dspace-1.2.1-source/src/org/dspace/app/webui/jsptag/ItemListTag.java	Thu Feb 10 13:05:47 2005
--- /dspace/orig-1.2.2/dspace-1.2.2-source/src/org/dspace/app/webui/jsptag/ItemListTag.java	Mon Mar 28 22:11:15 2005
***************
*** 1,9 ****
  /*
   * ItemListTag.java
   *
!  * Version: $Revision: 1.17 $
   *
!  * Date: $Date: 2004/12/22 17:48:39 $
   *
   * Copyright (c) 2002, Hewlett-Packard Company and Massachusetts
   * Institute of Technology.  All rights reserved.
--- 1,9 ----
  /*
   * ItemListTag.java
   *
!  * Version: $Revision: 1.20 $
   *
!  * Date: $Date: 2005/03/29 03:11:15 $
   *
   * Copyright (c) 2002, Hewlett-Packard Company and Massachusetts
   * Institute of Technology.  All rights reserved.
***************
*** 62,69 ****
  import java.io.InputStream;
  import java.io.UnsupportedEncodingException;
  
- import java.net.URLEncoder;
- 
  import java.sql.SQLException;
  
  import javax.imageio.ImageIO;
--- 62,67 ----
***************
*** 77,83 ****
   * Tag for display a list of items
   * 
   * @author Robert Tansley
!  * @version $Revision: 1.17 $
   */
  public class ItemListTag extends TagSupport
  {
--- 75,81 ----
   * Tag for display a list of items
   * 
   * @author Robert Tansley
!  * @version $Revision: 1.20 $
   */
  public class ItemListTag extends TagSupport
  {
***************
*** 496,502 ****
                                  .append(item.getHandle()).append("/").append(
                                          originalBitstream.getSequenceID())
                                  .append("/").append(
!                                         URLEncoder.encode(originalBitstream
                                                  .getName(),
                                                  Constants.DEFAULT_ENCODING));
                      }
--- 494,500 ----
                                  .append(item.getHandle()).append("/").append(
                                          originalBitstream.getSequenceID())
                                  .append("/").append(
!                                 		UIUtil.encodeBitstreamName(originalBitstream
                                                  .getName(),
                                                  Constants.DEFAULT_ENCODING));
                      }
***************
*** 510,516 ****
                      thumbLink.append("\"><img src=\"").append(
                              hrq.getContextPath()).append("/retrieve/").append(
                              thumbnailBitstream.getID()).append("/").append(
!                             URLEncoder.encode(thumbnailBitstream.getName(),
                                      Constants.DEFAULT_ENCODING)).append(
                              "\" alt=\"").append(thumbnailBitstream.getName())
                              .append("\" ").append(
--- 508,514 ----
                      thumbLink.append("\"><img src=\"").append(
                              hrq.getContextPath()).append("/retrieve/").append(
                              thumbnailBitstream.getID()).append("/").append(
!                             		UIUtil.encodeBitstreamName(thumbnailBitstream.getName(),
                                      Constants.DEFAULT_ENCODING)).append(
                              "\" alt=\"").append(thumbnailBitstream.getName())
                              .append("\" ").append(
