Index: org/apache/solr/client/solrj/embedded/JettySolrRunner.java
===================================================================
--- org/apache/solr/client/solrj/embedded/JettySolrRunner.java	(revision 885060)
+++ org/apache/solr/client/solrj/embedded/JettySolrRunner.java	(working copy)
@@ -17,22 +17,21 @@
 
 package org.apache.solr.client.solrj.embedded;
 
-import java.io.IOException;
-import java.io.InputStream;
-import java.net.URL;
-
-import javax.servlet.http.HttpServlet;
-import javax.servlet.http.HttpServletRequest;
-import javax.servlet.http.HttpServletResponse;
-
 import org.apache.solr.servlet.SolrDispatchFilter;
+import org.mortbay.jetty.Connector;
 import org.mortbay.jetty.Handler;
 import org.mortbay.jetty.Server;
-import org.mortbay.jetty.Connector;
 import org.mortbay.jetty.servlet.Context;
 import org.mortbay.jetty.servlet.FilterHolder;
 import org.mortbay.log.Logger;
 
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URL;
+
 /**
  * Run solr using jetty
  * 
@@ -55,6 +54,13 @@
     this.init( context, port );
     dispatchFilter.setInitParameter("solrconfig-filename", solrConfigFilename);
   }
+
+  public JettySolrRunner( String context, int port, String solrConfigFilename, String indexSchemaFilename)
+  {
+    this.init( context, port );
+    dispatchFilter.setInitParameter("solrconfig-filename", solrConfigFilename);
+    dispatchFilter.setInitParameter("schema-filename", indexSchemaFilename);
+  }
   
 //  public JettySolrRunner( String context, String home, String dataDir, int port, boolean log )
 //  {
Index: org/apache/solr/servlet/SolrDispatchFilter.java
===================================================================
--- org/apache/solr/servlet/SolrDispatchFilter.java	(revision 885060)
+++ org/apache/solr/servlet/SolrDispatchFilter.java	(working copy)
@@ -17,33 +17,30 @@
 
 package org.apache.solr.servlet;
 
-import java.io.IOException;
-import java.io.PrintWriter;
-import java.io.StringWriter;
-import java.io.ByteArrayInputStream;
-import java.util.Map;
-import java.util.WeakHashMap;
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
-
-import javax.servlet.Filter;
-import javax.servlet.FilterChain;
-import javax.servlet.FilterConfig;
-import javax.servlet.ServletException;
-import javax.servlet.ServletRequest;
-import javax.servlet.ServletResponse;
-import javax.servlet.http.HttpServletRequest;
-import javax.servlet.http.HttpServletResponse;
-
 import org.apache.solr.common.SolrException;
+import org.apache.solr.common.params.CommonParams;
 import org.apache.solr.common.util.NamedList;
 import org.apache.solr.common.util.SimpleOrderedMap;
-import org.apache.solr.common.params.CommonParams;
-import org.apache.solr.core.*;
+import org.apache.solr.core.Config;
+import org.apache.solr.core.CoreContainer;
+import org.apache.solr.core.SolrConfig;
+import org.apache.solr.core.SolrCore;
 import org.apache.solr.request.*;
 import org.apache.solr.servlet.cache.HttpCacheHeaderUtil;
 import org.apache.solr.servlet.cache.Method;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
+import javax.servlet.*;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.io.StringWriter;
+import java.util.Map;
+import java.util.WeakHashMap;
+
 /**
  * This filter looks at the incoming URL maps them to handlers defined in solrconfig.xml
  *
@@ -79,6 +76,7 @@
       // web.xml configuration
       this.pathPrefix = config.getInitParameter( "path-prefix" );
       init.setSolrConfigFilename(config.getInitParameter("solrconfig-filename"));
+      init.setIndexSchemaFilename(config.getInitParameter("schema-filename"));
 
       this.cores = init.initialize();
       abortOnConfigurationError = init.isAbortOnConfigurationError();
